components/pipefy/sources/common-polling.js

```diff
@@ -0,0 +1,28 @@
+const common = require("./common.js");
+
+module.exports = {
+  ...common,
+  props: {
+    ...common.props,
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 15,
+      },
+    },
+  },
+  async run() {
+    const cards = await this.pipefy.listCards(this.pipeId);
+    for (const edge of cards.edges) {
+      const { node } = edge;
+      const { due_date } = node;
+      const due = due_date ? new Date(due_date) : new Date();
+      if (!this.isCardRelevant(node, due)) continue;
+      this.$emit(node, {
+        id: this.getEmitId(node),
+        summary: node.title,
+        ts: due.getTime(),
```

Comment:
I don't think this is correct. `ts` is supposed to be the timestamp of the event itself, but in this case the timestamp of the event (i.e. the time at which an event was emitted) could have a different meaning depending on the content of the card and use case.

I think in this case the logic from `card-overdue` leaked itself out of its own module and into this common module (which is supposed to only contain logic related to polling the API for new records and processing them in an abstract way).

If we do something like this instead, then we can refine the behaviour for each use case:
```js
// common-polling.js
async run() {
  const cards = await this.pipefy.listCards(this.pipeId);
  for (const edge of cards.edges) {
    const { node } = edge;
    if (!this.isCardRelevant({ node, event })) continue;

    const meta = this.getMeta(node);
    this.$emit(node, meta);
  }
}
```

Then each event source would implement the specific logic. In the case of `card-overdue`, the due date would be in fact the timestamp (assuming that's the expected behaviour):
```js
// card-overdue.js
module.exports = {
  methods: {
    isCardRelevant({ node, event }) {
      // We take `event` as input since this is the
      // actual monotonic timestamp that should be used as reference.
      // This will avoid calling Date.now(), which will make the
      // event source deterministic.
      const { timestamp: eventTimestamp } = event;

      // To match Javascript's timestamp, since in Javascript epochs
      // are expressed in milliseconds instead of seconds
      const eventDate = eventTimestamp * 1000;
      
      const { due_date: dueDateIso } = node;
      const dueDate = Date.parse(dueDateIso);
      
      return (
        dueDate < eventDate &&
        !node.done
      );
    },
    getMeta(node) {
      const {
        id,
        title: summary,
        due_date: dueDate,
      } = node;
      const ts = Date.parse(dueDate) || Date.now();
      return {
        id,
        summary,
        ts,
      };
    },
  },
};
```

The other cases are much simpler I think, for example:
```js
// card-expired.js
module.exports = {
  methods: {
    isCardRelevant({ node }) {
      return (
        node.expired &&
        !node.done
      );
    },
    getMeta(node) {
      const {
        id: nodeId,
        title: summary,
        current_phase: { id: currentPhaseId },
      } = node;
      const id = `${nodeId}${currentPhaseId}`;
      const ts = Date.now();
      return {
        id,
        summary,
        ts,
      };
    },
  },
}; 
```